crypto/ts/quinTree.ts

```diff
@@ -0,0 +1,373 @@
+   * @param value The value to insert
+   */
+  insert(value: Leaf): void {
+    // update the node with this leaf
+    this.update(this.nextIndex, value);
+    this.nextIndex += 1;
+  }
+
+  /**
+   * Update a leaf at a given index
+   * @param index The index of the leaf to update
+   * @param value The value to update the leaf with
+   */
+  update(index: number, value: Leaf): void {
+    // Set the leaf value
+    this.setNode(index, value);
+
+    // Set the parent leaf value
+    // Get the parent indices
+    const parentIndices = this.calcParentIndices(index);
+
+    parentIndices.forEach((parentIndex) => {
+      const childIndices = this.calcChildIndices(parentIndex);
+
+      const elements: Leaf[] = [];
+      childIndices.forEach((childIndex) => {
+        elements.push(this.getNode(childIndex));
+      });
+      this.nodes[parentIndex] = this.hashFunc(elements);
+    });
+
+    this.root = this.nodes[this.numNodes - 1];
+  }
+
+  /**
+   * Calculate the indices of the leaves in the path to the root
+   * @param index The index of the leaf
+   * @returns The indices of the leaves in the path to the root
+   */
+  calcLeafIndices(index: number): number[] {
+    const indices = new Array<number>(this.depth);
+
+    let r = index;
+    for (let i = 0; i < this.depth; i += 1) {
+      indices[i] = r % this.arity;
+      r = Math.floor(r / this.arity);
+    }
+
+    return indices;
+  }
+
+  /**
+   * Generate a proof for a given leaf index
+   * @param index The index of the leaf to generate a proof for
+   * @returns The proof
+   */
+  genProof(index: number): IMerkleProof {
+    if (index < 0) {
+      throw new Error("The leaf index must be greater or equal to 0");
+    }
+
+    if (index >= this.capacity) {
+      throw new Error("+The leaf index must be less than the tree capacity");
+    }
+
+    const pathElements: bigint[][] = [];
+    const indices = this.calcLeafIndices(index);
+
+    // Calculate path elements
+    let leafIndex = index;
+    let offset = 0;
+
+    for (let i = 0; i < this.depth; i += 1) {
+      const elements: bigint[] = [];
+      const start = leafIndex - (leafIndex % this.arity) + offset;
+
+      for (let j = 0; j < this.arity; j += 1) {
+        if (j !== indices[i]) {
+          const node = this.getNode(start + j);
+          elements.push(node);
+        }
+      }
+
+      pathElements.push(elements);
+      leafIndex = Math.floor(leafIndex / this.arity);
+      offset += this.arity ** (this.depth - i);
+    }
+
+    return {
+      pathElements,
+      pathIndices: indices,
+      root: this.root,
+      leaf: this.getNode(index),
+    };
+  }
+
+  /**
+   * Generates a Merkle proof from a subroot to the root.
+   * @param startIndex The index of the first leaf
+   * @param endIndex The index of the last leaf
+   * @returns The Merkle proof
+   */
+  genSubrootProof(
+    // inclusive
+    startIndex: number,
+    // exclusive
+    endIndex: number,
+  ): IMerkleProof {
```

Comment:
I saw we check if the `index` is negative in the `genProof` function, should we check it here too?